### ISSUE TITLE
`cargo build` of root crate staticlib fails if FS doesn't support hardlinks

### ISSUE THREAD

**Issue by cbiffle** [rust-lang/cargo#3098]
**tl;dr:** Building a static library root crate will fail at a late stage if the filesystem containing `target/` does not support hardlinks.  Falling back to `copy` would probably be an acceptable workaround.

Cargo tries to replicate products of the root crate from `deps/` into a higher level of `target/`.  The current code was introduced in 0863469, and uses `fs::hard_link` [here](https://github.com/rust-lang/cargo/blob/f4a23e7929fdc802a78671e9637a18ae1ecb566a/src/cargo/ops/cargo_rustc/mod.rs#L319).  This fails if the filesystem containing `target/` doesn't support hardlinks.

This isn't specific to legacy filesystems like FAT32.  In particular, it is currently hard to develop a library using Cargo inside a [Vagrant](https://www.vagrantup.com/) environment, because Vagrant maps the source tree into the VM (by default) using VirtualBox's shared folder filesystem... which doesn't support hardlinks.

Typical build output:

```
vagrant@vagrant-ubuntu-trusty-64:/vagrant$ cargo build
   Compiling emb1 v0.1.0 (file:///vagrant)
error: failed to link `/vagrant/target/thumbv7em-none-eabi/debug/deps/libemb1.a` to `/vagrant/target/thumbv7em-none-eabi/debug/libemb1.a`
```

With verbose we can see that the hardlink call is failing with "operation not permitted:"

```
vagrant@vagrant-ubuntu-trusty-64:/vagrant$ cargo build --verbose
       Fresh rust-libcore v0.0.3
   Compiling emb1 v0.1.0 (file:///vagrant)
     Running `rustc src/emb1.rs --crate-name emb1 --crate-type staticlib -C panic=abort -g -C metadata=d5552433d0098705 --out-dir /vagrant/target/thumbv7em-none-eabi/debug/deps --emit=dep-info,link --target thumbv7em-none-eabi -C linker=arm-none-eabi-gcc -L dependency=/vagrant/target/thumbv7em-none-eabi/debug/deps --extern core=/vagrant/target/thumbv7em-none-eabi/debug/deps/libcore-882297ed0c39d543.rlib`
error: failed to link `/vagrant/target/thumbv7em-none-eabi/debug/deps/libemb1.a` to `/vagrant/target/thumbv7em-none-eabi/debug/libemb1.a`

Caused by:
  Operation not permitted (os error 1)
```

From what I can see, `fs::hard_link` is being used as an optimization; it's not clear that hard linking is necessary for correctness, and it would probably be okay to fall back to copying the file.  (This has been suggested before back in [this comment](https://github.com/rust-lang/cargo/issues/1924#issuecomment-198447385) on #1924.)

Using hardlinks unconditionally has other problems; see #800.  #2956 didn't include specifics about the build environment (e.g. whether it was in a VM) but may be caused by FS hardlink support.

**Comment by alexcrichton** [comment]
Ah yeah, it's intended that these sorts of links all have copy fallabacks. Should be easy to add!

**Comment by dgellow** [comment]
Still an issue? The PR #3118 has been merged

**Comment by alexcrichton** [comment]
Indeed!
